core/iwasm/aot/aot_loader.c

```diff
@@ -3232,7 +3236,15 @@ aot_load_from_aot_file(const uint8 *buf, uint32 size, char *error_buf,
     if (!module)
         return NULL;
 
-    if (!load(buf, size, module, error_buf, error_buf_size)) {
+#if defined(__APPLE__) && defined(__arm64__)
+    pthread_jit_write_protect_np(false); // Make memory writable
+#endif
+    bool ret = load(buf, size, module, error_buf, error_buf_size);
+#if defined(__APPLE__) && defined(__arm64__)
+    pthread_jit_write_protect_np(true); // Make memory executable
+    sys_icache_invalidate(module->code, module->code_size);
+#endif
```

Comment:
We usually don't directly call the libc functions in core/iwasm, instead, we encapsulated them in the platform layer and call them. How about adding API in core/shared/platform, e.g. `void os_thread_jit_write_protect_np(bool enabled)`, in:

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/include/platform_api_vmcore.h#L83

and:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_thread.c#L420

and call it here.

and how about wrappring sys_icache_invalidate to os_icache_flush? like os_dcache_flush:

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/include/platform_api_vmcore.h#L144

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/common/posix/posix_memmap.c#L253C1-L253C1

And had better set memory executable after ret is true. Reference code:

```C
#if defined(__APPLE__) && defined(__arm64__)
    os_thread_jit_write_protect_np(false); /* Make memory writable */
#endif

    if (!load(buf, size, module, error_buf, error_buf_size)) {
        aot_unload(module);
        return NULL;
    }

#if defined(__APPLE__) && defined(__arm64__)
    if (module->code) {
        os_thread_jit_write_protect_np(true); /* Make memory executable */
        os_icache_flush(module->code, module->code_size);
    }
#endif
```

Reply:
Thanks, I should have known.
The ifdef should be inside the platform-specific code I think, and no need for `if (module->code)` since `load` would have returned false if that was NULL.
```c
os_thread_jit_write_protect_np(false); /* Make memory writable */
if (!load(buf, size, module, error_buf, error_buf_size)) {
    aot_unload(module);
    return NULL;
}
os_thread_jit_write_protect_np(true); /* Make memory executable */
os_icache_flush(module->code, module->code_size);
```

Reply:
OK, `module->code` may be NULL if the module hasn't function bytecodes, but I think the pthread_jit_write_protect_np should have handle that case.